File: lib/services/if_engine/ink_engine_connector.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:flutter_js/flutter_js.dart';
import 'if_engine_connector.dart';

/// ink engine implementation of [IfEngineConnector].
///
/// Uses inkjs (v2.4.0) running inside flutter_js for both compilation and
/// runtime execution, entirely client-side.
///
/// Requires [initialize] to be called before any other method.
/// See docs/Ink Engine Connector.md for the full specification.
/// Implemented in: lib/services/if_engine/ink_engine_connector.dart
class InkEngineConnector implements IfEngineConnector {
  late JavascriptRuntime _jsRuntime;
  bool _initialized = false;

  /// Loads the inkjs library from assets into the JS runtime.
  /// Must be called before any other method.
  Future<void> initialize() async {
    if (_initialized) return;
    _jsRuntime = getJavascriptRuntime();
    final inkJs = await rootBundle.loadString('assets/ink-full.js');
    _jsRuntime.evaluate(inkJs);
    // Create a global storage object for the ink story
    _jsRuntime.evaluate('var _inkStory = null;');
    _initialized = true;
  }

  void _ensureInitialized() {
    if (!_initialized) {
      throw StateError('InkEngineConnector.initialize() must be called first');
    }
  }

  @override
  String getEngineName() => 'ink';

  @override
  String getFileExtension() => '.ink.json';

  @override
  String getScriptPrompt() => '''
ink scripts use a specific syntax for interactive narrative. Key requirements:
- IMPORTANT: The script MUST start with a divert to the opening knot (e.g., -> opening) at the very top of the file, BEFORE any knot declarations. Without this, the story will not run.
- Knots are declared with === knot_name ===
- Stitches within knots use = stitch_name
- Choices use * for one-time and + for sticky choices
- Choice text in brackets [] is not shown after selection
- Diverts use -> knot_name or -> knot_name.stitch_name
- Variables are declared with VAR name = value and modified with ~ name = value
- Use { condition: text } for conditional text
- Use { variable } to print variable values inline
- Use <> for glue to join text across lines without whitespace
- End threads with -> DONE and end the story with -> END
- Comments use // for single line and /* */ for multi-line
- External functions and INCLUDE are not supported in Z-Forge

CRITICAL - Narrative vs. Choices:
- ONLY player actions/decisions should be choices (lines starting with * or +)
- Narrative text, NPC dialogue, and scene descriptions must be plain text (NO * or + prefix)
- WRONG: * "Hello!" said the dragon. (This makes NPC dialogue a clickable choice!)
- RIGHT: "Hello!" said the dragon.
         * [Greet the dragon] "Hello to you too!"
- WRONG: * Choose your path:  (This makes instructions a choice!)
- RIGHT: Choose your path:
         * [Go left] -> left_path
         * [Go right] -> right_path

Script structure example:
VAR trust = 0

-> opening

=== opening ===
The forest stretched endlessly before you. A wise owl perched nearby.
"Welcome, traveler," the owl hooted softly.
* [Ask for directions] 
    "Which way to the village?" you ask.
    -> village_directions
* [Ignore the owl]
    You walk past without a word.
    -> forest_path

=== village_directions ===
The owl ruffles its feathers thoughtfully.
"Follow the stream north," it advises.
* [Thank the owl] -> thank_owl
* [Ask another question] -> more_questions

Common patterns:
- Branching: Use choices leading to different knots
- Loops: Use sticky choices (+) for repeatable options
- State tracking: Use VAR for counters, flags, and relationships
- Conditional choices: Use { condition } before * or + to show choices conditionally
''';

  @override
  Future<BuildResult> build(String script) async {
    _ensureInitialized();
    final escapedScript = _escapeForJs(script);
    final result = _jsRuntime.evaluate('''
      (function() {
        try {
          var compiler = new inkjs.Compiler("$escapedScript");
          var story = compiler.Compile();
          if (compiler.errors && compiler.errors.length > 0) {
            return JSON.stringify({
              success: false,
              errors: compiler.errors,
              warnings: compiler.warnings || []
            });
          }
          return JSON.stringify({
            success: true,
            json: story.ToJson(),
            warnings: compiler.warnings || []
          });
        } catch (e) {
          return JSON.stringify({
            success: false,
            errors: [e.toString()],
            warnings: []
          });
        }
      })()
    ''');
    final data = jsonDecode(result.stringResult) as Map<String, dynamic>;
    if (data['success'] == true) {
      return BuildResult(
        output: Uint8List.fromList(utf8.encode(data['json'] as String)),
        warnings: List<String>.from(data['warnings'] ?? []),
        errors: [],
      );
    } else {
      return BuildResult(
        output: null,
        warnings: List<String>.from(data['warnings'] ?? []),
        errors: List<String>.from(data['errors'] ?? []),
      );
    }
  }

  @override
  Future<String> startExperience(Uint8List compiledData) async {
    _ensureInitialized();
    final jsonString = utf8.decode(compiledData);
    debugPrint('[InkEngine] startExperience: JSON length=${jsonString.length}');
    final escapedJson = _escapeForJs(jsonString);
    final createResult = _jsRuntime.evaluate('''
      (function() {
        try {
          _inkStory = new inkjs.Story("$escapedJson");
          return JSON.stringify({ success: true });
        } catch (e) {
          return JSON.stringify({ success: false, error: e.toString() });
        }
      })()
    ''');
    debugPrint('[InkEngine] Create story result: ${createResult.stringResult}');
    final createData = jsonDecode(createResult.stringResult) as Map<String, dynamic>;
    if (createData['success'] != true) {
      throw Exception('Failed to create ink story: ${createData['error']}');
    }
    
    // Try to continue story; if no content, try jumping to common start knots
    final tryResult = _jsRuntime.evaluate('''
      (function() {
        var story = _inkStory;
        var text = "";
        
        // First, try to get content from where the story starts
        while (story.canContinue) {
          text += story.Continue();
        }
        
        // If we got content or choices, we're good
        if (text.trim().length > 0 || story.currentChoices.length > 0) {
          return JSON.stringify({ success: true, text: text, method: "direct" });
        }
        
        // No content at root - try jumping to common starting knot names
        var startKnots = ["opening", "start", "begin", "intro", "main"];
        for (var i = 0; i < startKnots.length; i++) {
          try {
            var container = story.KnotContainerWithName(startKnots[i]);
            if (container) {
              story.ChoosePathString(startKnots[i]);
              text = "";
              while (story.canContinue) {
                text += story.Continue();
              }
              if (text.trim().length > 0 || story.currentChoices.length > 0) {
                return JSON.stringify({ success: true, text: text, method: "jumped", knot: startKnots[i] });
              }
            }
          } catch (e) {
            // Continue trying other knots
          }
        }
        
        return JSON.stringify({ success: false, error: "No playable content found in story" });
      })()
    ''');
    debugPrint('[InkEngine] Try result: ${tryResult.stringResult}');
    
    final tryData = jsonDecode(tryResult.stringResult) as Map<String, dynamic>;
    if (tryData['success'] != true) {
      throw Exception(tryData['error'] ?? 'Failed to start story');
    }
    return (tryData['text'] as String? ?? '').trim();
  }

  @override
  Future<ActionResult> takeAction(String input) async {
    _ensureInitialized();
    final choiceIndex = int.tryParse(input);
    if (choiceIndex == null) {
      throw ArgumentError('ink requires numeric choice index, got: $input');
    }
    _jsRuntime.evaluate('''
      _inkStory.ChooseChoiceIndex($choiceIndex);
    ''');
    final text = _continueStory();
    final stateResult = _jsRuntime.evaluate('''
      (function() {
        var story = _inkStory;
        return JSON.stringify({
          choices: story.currentChoices.map(function(c) { return c.text; }),
          isComplete: !story.canContinue && story.currentChoices.length === 0
        });
      })()
    ''');
    final stateData =
        jsonDecode(stateResult.stringResult) as Map<String, dynamic>;
    final choices = List<String>.from(stateData['choices'] ?? []);
    return ActionResult(
      text: text,
      choices: choices.isEmpty ? null : choices,
      isComplete: stateData['isComplete'] as bool,
    );
  }

  @override
  Future<Uint8List> saveState() async {
    _ensureInitialized();
    final result = _jsRuntime.evaluate('''
      JSON.stringify(_inkStory.state.ToJson());
    ''');
    return Uint8List.fromList(utf8.encode(result.stringResult));
  }

  @override
  Future<String> restoreState(Uint8List savedState) async {
    _ensureInitialized();
    final stateJson = utf8.decode(savedState);
    final escapedState = _escapeForJs(stateJson);
    _jsRuntime.evaluate('''
      _inkStory.state.LoadJson("$escapedState");
    ''');
    return _continueStory();
  }

  /// Returns available choices without advancing the story.
  Future<List<String>> getCurrentChoices() async {
    _ensureInitialized();
    final result = _jsRuntime.evaluate('''
      (function() {
        try {
          if (!_inkStory) {
            return JSON.stringify({ success: false, error: "Story not initialized" });
          }
          var choices = _inkStory.currentChoices.map(function(c) { return c.text; });
          return JSON.stringify({ success: true, choices: choices });
        } catch (e) {
          return JSON.stringify({ success: false, error: e.toString() });
        }
      })()
    ''');
    final data = jsonDecode(result.stringResult) as Map<String, dynamic>;
    if (data['success'] != true) {
      throw Exception('Failed to get choices: ${data['error']}');
    }
    return List<String>.from(data['choices'] as List);
  }

  /// Reads the current value of a story variable.
  Future<dynamic> getVariable(String name) async {
    _ensureInitialized();
    final escapedName = _escapeForJs(name);
    final result = _jsRuntime.evaluate('''
      JSON.stringify(_inkStory.variablesState["$escapedName"])
    ''');
    return jsonDecode(result.stringResult);
  }

  /// Sets a story variable to [value].
  Future<void> setVariable(String name, dynamic value) async {
    _ensureInitialized();
    final escapedName = _escapeForJs(name);
    final valueJson = jsonEncode(value);
    _jsRuntime.evaluate('''
      _inkStory.variablesState["$escapedName"] = $valueJson;
    ''');
  }

  /// Releases the JS runtime. Call when done with this connector.
  void dispose() {
    if (_initialized) {
      _jsRuntime.dispose();
      _initialized = false;
    }
  }

  String _continueStory() {
    final result = _jsRuntime.evaluate('''
      (function() {
        try {
          var story = _inkStory;
          if (!story) {
            return JSON.stringify({ success: false, error: "Story not initialized" });
          }
          var text = "";
          while (story.canContinue) {
            text += story.Continue();
          }
          return JSON.stringify({ success: true, text: text, canContinue: story.canContinue, choicesCount: story.currentChoices.length });
        } catch (e) {
          return JSON.stringify({ success: false, error: e.toString() });
        }
      })()
    ''');
    debugPrint('[InkEngine] _continueStory result: ${result.stringResult}');
    final data = jsonDecode(result.stringResult) as Map<String, dynamic>;
    if (data['success'] != true) {
      throw Exception('Failed to continue story: ${data['error']}');
    }
    return (data['text'] as String? ?? '').trim();
  }

  String _escapeForJs(String input) {
    return input
        .replaceAll('\\', '\\\\')
        .replaceAll('"', '\\"')
        .replaceAll('\n', '\\n')
        .replaceAll('\r', '\\r')
        .replaceAll('\t', '\\t');
  }
}

File: src/zforge/services/if_engine/ink_engine_connector.py
"""Ink Engine Connector implementation.

Uses py-mini-racer + inkjs for in-process ink compilation and runtime.
Implements: src/zforge/services/if_engine/ink_engine_connector.py per
docs/Ink Engine Connector.md and docs/IF Engine Abstraction Layer.md.

Note: restore_state() returns ActionResult per the ABC spec in
docs/IF Engine Abstraction Layer.md (not str as shown in the
Ink Engine Connector doc example).
"""

from __future__ import annotations

import json
from pathlib import Path

import quickjs

from zforge.models.results import ActionResult, BuildResult
from zforge.services.if_engine.if_engine_connector import IfEngineConnector


_ASSETS_DIR = Path(__file__).resolve().parent.parent.parent.parent.parent / "assets"


class InkEngineConnector(IfEngineConnector):
    """IF engine connector for ink via quickjs + inkjs."""

    def __init__(self) -> None:
        self._ctx: quickjs.Context | None = None

    async def initialize(self) -> None:
        """Must be called before any other method."""
        if self._ctx is not None:
            return
        ink_js = (_ASSETS_DIR / "ink-full.js").read_text(encoding="utf-8")
        ctx = quickjs.Context()
        ctx.eval(ink_js)
        self._ctx = ctx

    def _ensure_initialized(self) -> None:
        if self._ctx is None:
            raise RuntimeError(
                "InkEngineConnector.initialize() must be called first"
            )

    def get_engine_name(self) -> str:
        return "ink"

    def get_file_extension(self) -> str:
        return ".ink.json"

    def get_script_prompt(self) -> str:
        return """
ink scripts use a specific syntax for interactive narrative. Key requirements:
- IMPORTANT: The script MUST start with a divert to the opening knot (e.g., -> opening) at the very top of the file, BEFORE any knot declarations. Without this, the story will not run.
- Knots are declared with === knot_name ===
- Stitches within knots use = stitch_name
- Choices use * for one-time and + for sticky choices
- Choice text in brackets [] is not shown after selection
- Diverts use -> knot_name or -> knot_name.stitch_name
- Variables are declared with VAR name = value and modified with ~ name = value
- CRITICAL: ~ assignments MUST be on their own dedicated line — NEVER on the same line as narrative text or dialogue. WRONG: `"She smiled." ~ courage = true` (the assignment becomes visible text!). RIGHT: put `~ courage = true` on its own separate line before or after the narrative.
- Use { condition: text } for conditional text
- Use { variable } to print variable values inline
- Use <> for glue to join text across lines without whitespace
- End threads with -> DONE and end the story with -> END
- Comments use // for single line and /* */ for multi-line
- External functions and INCLUDE are not supported in Z-Forge

CRITICAL - Narrative vs. Choices:
- ONLY player actions/decisions should be choices (lines starting with * or +)
- Narrative text, NPC dialogue, and scene descriptions must be plain text (NO * or + prefix)
- WRONG: * "Hello!" said the dragon. (This makes NPC dialogue a clickable choice!)
- RIGHT: "Hello!" said the dragon.
         * [Greet the dragon] "Hello to you too!"
- WRONG: * Choose your path:  (This makes instructions a choice!)
- RIGHT: Choose your path:
         * [Go left] -> left_path
         * [Go right] -> right_path

Script structure example:
VAR trust = 0

-> opening

=== opening ===
The forest stretched endlessly before you. A wise owl perched nearby.
"Welcome, traveler," the owl hooted softly.
* [Ask for directions]
    "Which way to the village?" you ask.
    -> village_directions
* [Ignore the owl]
    You walk past without a word.
    -> forest_path

=== village_directions ===
The owl ruffles its feathers thoughtfully.
"Follow the stream north," it advises.
* [Thank the owl] -> thank_owl
* [Ask another question] -> more_questions

Common patterns:
- Branching: Use choices leading to different knots
- Loops: Use sticky choices (+) for repeatable options
- State tracking: Use VAR for counters, flags, and relationships
- Conditional choices: Use { condition } before * or + to show choices conditionally
"""

    async def build(self, script: str) -> BuildResult:
        self._ensure_initialized()
        escaped = json.dumps(script)
        result = self._ctx.eval(f"""
            (function() {{
                try {{
                    var compiler = new inkjs.Compiler({escaped});
                    var story = compiler.Compile();
                    return JSON.stringify({{
                        success: true,
                        json: story.ToJson(),
                        warnings: compiler.warnings || []
                    }});
                }} catch (e) {{
                    return JSON.stringify({{
                        success: false,
                        errors: [e.toString()],
                        warnings: []
                    }});
                }}
            }})()
        """)
        data = json.loads(result)
        if data["success"]:
            return BuildResult(
                output=data["json"].encode("utf-8"),
                warnings=data.get("warnings", []),
                errors=[],
            )
        return BuildResult(
            output=None,
            warnings=data.get("warnings", []),
            errors=data.get("errors", []),
        )

    async def start_experience(self, compiled_data: bytes) -> str:
        self._ensure_initialized()
        json_str = compiled_data.decode("utf-8")
        # inkjs.Story expects a parsed JSON object. Ensure we pass a JS object
        # by calling JSON.parse(...) on the compiled JSON string.
        self._ctx.eval(
            f"globalThis._inkStory = new inkjs.Story(JSON.parse({json.dumps(json_str)}));"
        )
        # Fault-tolerance: if the story starts in a terminal state (no content
        # and no choices), or if it canContinue but produces no text (a "ghost" start),
        # the generated script likely has a missing or broken root divert.
        # Attempt recovery by seeking the first named knot.
        recovery = self._ctx.eval(f"""
            (function() {{
                var story = globalThis._inkStory;
                var inkJson = JSON.parse({json.dumps(json_str)});
                var logs = [];
                
                var needsRecovery = (!story.canContinue && story.currentChoices.length === 0);
                
                if (!needsRecovery && story.canContinue) {{
                    var state = story.state.ToJson();
                    var text = "";
                    while (story.canContinue) {{ text += story.Continue(); }}
                    var producedNothing = text.trim().length === 0;
                    story.state.LoadJson(state);
                    if (producedNothing && story.currentChoices.length === 0) {{
                        needsRecovery = true;
                    }}
                }}

                if (needsRecovery) {{
                    logs.push("Recovery triggered");
                    var excluded = ["inkVersion", "root", "listDefs"];
                    
                    // Helper to check an object for knots
                    function findInObj(obj) {{
                        return Object.keys(obj).find(function(k) {{
                            return (excluded.indexOf(k) === -1 && k.indexOf("g-") !== 0 && k !== "#f");
                        }});
                    }}

                    // Search top level
                    var entry = findInObj(inkJson);
                    
                    // Search in root array elements (nested knots)
                    if (!entry && inkJson.root && Array.isArray(inkJson.root)) {{
                        for (var i = 0; i < inkJson.root.length; i++) {{
                            var item = inkJson.root[i];
                            if (item && typeof item === "object" && !Array.isArray(item)) {{
                                entry = findInObj(item);
                                if (entry) {{
                                    logs.push("Found nested entry: " + entry + " in root[" + i + "]");
                                    break;
                                }}
                            }}
                        }}
                    }}

                    if (entry) {{
                        try {{
                            story.ChoosePathString(entry);
                            return "recovered:" + entry + " | logs: " + logs.join("; ");
                        }} catch (e) {{
                            return "recovery-failed:" + entry + " " + e.toString();
                        }}
                    }}
                    return "no-entry-found | logs: " + logs.join("; ");
                }}
                return "ok";
            }})()
        """)
        if recovery != "ok":
            print("InkEngineConnector.start_experience: recovery attempt ->", recovery)
        # Diagnostic: inspect initial story state and currentChoices to aid debugging.
        try:
            diag = self._ctx.eval(
                "JSON.stringify({canContinue: globalThis._inkStory.canContinue, choices: globalThis._inkStory.currentChoices.map(function(c){return c.text;})})"
            )
            print("InkEngineConnector.start_experience: story diag ->", diag)
        except Exception as e:
            print("InkEngineConnector.start_experience: diag eval failed:", e)
        text = self._continue_story()
        print("InkEngineConnector.start_experience: continued text type=", type(text), "repr=", repr(text))
        return text

    def _continue_story(self) -> str:
        result = self._ctx.eval("""
            (function() {
                var story = globalThis._inkStory;
                var text = "";
                while (story.canContinue) { text += story.Continue(); }
                return text;
            })()
        """)
        print("InkEngineConnector._continue_story: raw result repr=", result)
        # Return the raw string — caller may inspect/strip as needed.
        try:
            return result
        except Exception:
            return str(result)

    async def take_action(self, input: str) -> ActionResult:
        self._ensure_initialized()
        choice_index = int(input)
        self._ctx.eval(
            f"globalThis._inkStory.ChooseChoiceIndex({choice_index});"
        )
        text = self._continue_story()
        state = json.loads(self._ctx.eval("""
            JSON.stringify({
                choices: globalThis._inkStory.currentChoices.map(function(c){return c.text;}),
                isComplete: !globalThis._inkStory.canContinue && globalThis._inkStory.currentChoices.length === 0
            })
        """))
        choices = state["choices"] or None
        return ActionResult(
            text=text, choices=choices, is_complete=state["isComplete"]
        )

    async def save_state(self) -> bytes:
        self._ensure_initialized()
        result = self._ctx.eval(
            "JSON.stringify(globalThis._inkStory.state.ToJson())"
        )
        return result.encode("utf-8")

    async def restore_state(self, saved_state: bytes) -> ActionResult:
        """Restore state and return ActionResult (per ABC spec)."""
        self._ensure_initialized()
        state_json = saved_state.decode("utf-8")
        self._ctx.eval(
            f"globalThis._inkStory.state.LoadJson({json.dumps(state_json)});"
        )
        text = self._continue_story()
        state = json.loads(self._ctx.eval("""
            JSON.stringify({
                choices: globalThis._inkStory.currentChoices.map(function(c){return c.text;}),
                isComplete: !globalThis._inkStory.canContinue && globalThis._inkStory.currentChoices.length === 0
            })
        """))
        choices = state["choices"] or None
        return ActionResult(
            text=text, choices=choices, is_complete=state["isComplete"]
        )

    # --- Helper methods (not part of abstract interface) ---

    async def get_current_choices(self) -> list[str]:
        """Get current available choices without making a selection."""
        self._ensure_initialized()
        return json.loads(self._ctx.eval(
            "JSON.stringify(globalThis._inkStory.currentChoices.map(function(c){return c.text;}))"
        ))

    async def get_variable(self, name: str):
        """Get current value of a story variable."""
        self._ensure_initialized()
        return json.loads(self._ctx.eval(
            f"JSON.stringify(globalThis._inkStory.variablesState[{json.dumps(name)}])"
        ))

    async def set_variable(self, name: str, value) -> None:
        """Set a story variable."""
        self._ensure_initialized()
        self._ctx.eval(
            f"globalThis._inkStory.variablesState[{json.dumps(name)}] = {json.dumps(value)};"
        )

    def dispose(self) -> None:
        """Release the JS runtime."""
        self._ctx = None
